sdk/src/extras/disk.rs

```diff
@@ -0,0 +1,711 @@
+use std::io::Read;
+use std::path::{Path, PathBuf};
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::{mpsc, oneshot, watch};
+
+impl ClientExtraDiskExt for Client {
+    fn disk_import(&self) -> builder::DiskImport {
+        builder::DiskImport::new(self)
+    }
+}
+
+pub mod builder {
+    use super::{DiskInfo, ImageInfo};
+    use crate::types::{self, Name, NameOrId};
+    use crate::{Client, Error};
+
+    use anyhow::Result;
+    use std::future::Future;
+    use std::sync::atomic::{AtomicBool, AtomicU64};
+    use std::sync::Arc;
+    use tokio::sync::{oneshot, watch};
+
+    pub struct DiskImport<'a> {
+        client: &'a Client,
+        project: Result<NameOrId, String>,
+        description: Result<String, String>,
+        disk: Result<Name, String>,
+        disk_info: Result<DiskInfo, String>,
+        image_info: Result<Option<ImageInfo>, String>,
+    }
+
+    impl<'a> DiskImport<'a> {
+        pub fn new(client: &'a Client) -> Self {
+            Self {
+                client,
+                project: Err("project was not initialized".to_string()),
+                description: Err("description was not initialized".to_string()),
+                disk: Err("disk was not initialized".to_string()),
+                disk_info: Err("disk_info was not initialized".to_string()),
+                image_info: Ok(None),
+            }
+        }
+
+        pub fn project<V>(mut self, value: V) -> Self
+        where
+            V: std::convert::TryInto<NameOrId>,
+        {
+            self.project = value
+                .try_into()
+                .map_err(|_| "conversion to `NameOrId` for project failed".to_string());
+            self
+        }
+
+        pub fn description<V>(mut self, value: V) -> Self
+        where
+            V: std::convert::TryInto<String>,
+        {
+            self.description = value
+                .try_into()
+                .map_err(|_| "conversion to `String` for description failed".to_string());
+            self
+        }
+
+        pub fn disk<V>(mut self, value: V) -> Self
+        where
+            V: std::convert::TryInto<Name>,
+        {
+            self.disk = value
+                .try_into()
+                .map_err(|_| "conversion to `Name` for disk failed".to_string());
+            self
+        }
+
+        pub fn disk_info(mut self, value: DiskInfo) -> Self {
+            self.disk_info = Ok(value);
+            self
+        }
+
+        pub fn image_info(mut self, value: ImageInfo) -> Self {
+            self.image_info = Ok(Some(value));
+            self
+        }
+
+        /// Return a `Future` for the disk creation and a `DiskImportHandle` which
+        /// can be used to track upload progress and cancel the job.
+        pub fn execute(
+            self,
+            thread_ct: Option<usize>,
+        ) -> Result<
+            (
+                impl Future<Output = Result<()>> + 'a,
```

Comment:
I still need to read through everything, but are we able to returned a typed error instead of an `anyhow` error? I think it is fine at the CLI level to return them, but `anyhow` errors are quite difficult to handle when getting them back from an sdk.

Reply:
Good call, I've moved to a custom error type.